modules/auxiliary/gather/ldap_esc_vulnerable_cert_finder.rb

```diff
@@ -315,6 +353,142 @@ def find_esc3_vuln_cert_templates
     query_ldap_server_certificates(esc3_template_2_raw_filter, 'ESC3_TEMPLATE_2')
   end
 
+  def find_esc4_vuln_cert_templates
+    # Determine who we are authenticating with. Retrieve the username and user SID
+    whoami_response = ''
+    begin
+      whoami_response = @ldap.ldapwhoami
+    rescue Net::LDAP::Error => e
+      print_warning("The module failed to run the ldapwhoami command, ESC4 detection can't continue. Error was: #{e.class}: #{e.message}.")
+      return
+    end
+
+    if whoami_response.empty?
+      print_error("Unable to retrieve the username using ldapwhoami, ESC4 detection can't continue")
+      return
+    end
+
+    sam_account_name = whoami_response.split('\\')[1]
+    user_raw_filter = "(sAMAccountName=#{sam_account_name})"
+    attributes = ['objectSID', 'DN']
+    our_user = query_ldap_server(user_raw_filter, attributes)&.first
+    if our_user.nil?
+      print_warning("Unable to determine the User SID for #{sam_account_name}, ESC4 detection can't continue")
+      return
+    else
+      user_sid = Rex::Proto::MsDtyp::MsDtypSid.read(our_user[:objectsid].first).value
+    end
+
+    # Determine if we are a User or a Computer
+    fiter_user_or_computer = "(objectSid=#{user_sid})"
+    attributes = ['objectClass']
+    response_computer_or_user = query_ldap_server(fiter_user_or_computer, attributes)&.first
+
+    if response_computer_or_user.nil?
+      print_warning("Unable to determine the if SID belongs to a user account or a computer account, ESC4 detection can't continue")
+      return
+    end
+
+    if response_computer_or_user[:objectclass].include?('computer')
+      computer_or_user = 'computer'
+    else
+      computer_or_user = 'user'
+    end
+
+    # Get the domain SID (from the user_sid which is domain_sid + rid)
+    domain_sid = ''
+    if user_sid =~ /(S-[0-9]*-[0-9]*-[0-9]*-[0-9]*-[0-9]*-[0-9]*)/
+      domain_sid = Regexp.last_match(1)
+    else
+      print_warning("Unable to derive the Domain SID from the User SID, ESC4 detection can't continue")
+      return
+    end
+
+    user_groups = []
+    if computer_or_user == 'computer'
+      user_groups << "#{domain_sid}-#{Rex::Proto::Secauthz::WellKnownSids::DOMAIN_GROUP_RID_COMPUTERS}"
+    else
+      user_groups << "#{domain_sid}-#{Rex::Proto::Secauthz::WellKnownSids::DOMAIN_GROUP_RID_USERS}"
+      user_groups << Rex::Proto::Secauthz::WellKnownSids::SECURITY_AUTHENTICATED_USER_SID
+    end
```

Comment:
Have we tested if computer accounts are members of the `SECURITY_AUTHENTICATED_USER_SID` group? It'd also be worth testing if we should be adding `SECURITY_ANONYMOUS_LOGON_SID` to the list in either of these cases.

Reply:
Computer accounts are members of `SECURITY_AUTHENTICATED_USER_SID`. I'll need to make some changes. 

> A group that includes all users and computers with identities that have been authenticated. Authenticated Users doesn't include Guest even if the Guest account has a password.
This group includes authenticated security principals from any trusted domain, not only the current domain.

Anonymous Logon is defined as:
>A user who has connected to the computer without supplying a user name and password.
 
As it's written ESC4 detection depends on having a username defined although I'll do some testing here as well.